diff --git a/case_study/RiscvPmp/Base.v b/case_study/RiscvPmp/Base.v
index 45fc015..bb34287 100644
--- a/case_study/RiscvPmp/Base.v
+++ b/case_study/RiscvPmp/Base.v
@@ -166,6 +166,7 @@ Inductive PmpAddrMatch : Set :=
 | RISCV_SUB
 | RISCV_SRA
+| RISCV_ADD'
 .
 
 (* IOP are the I-type opcodes. *)
 Inductive IOP : Set :=
@@ -490,7 +491,7 @@ Section Finite.
 
   #[export,program] Instance ROP_finite :
     Finite ROP :=
-    {| enum := [RISCV_ADD;RISCV_SUB;RISCV_SLT;RISCV_SLTU;RISCV_SLL;RISCV_SRL;RISCV_SRA;RISCV_AND;RISCV_OR;RISCV_XOR] |}.
+    {| enum := [RISCV_ADD;RISCV_SUB;RISCV_SLT;RISCV_SLTU;RISCV_SLL;RISCV_SRL;RISCV_SRA;RISCV_AND;RISCV_OR;RISCV_XOR;RISCV_ADD'] |}.
 
   #[export,program] Instance IOP_finite :
     Finite IOP :=
diff --git a/case_study/RiscvPmp/Machine.v b/case_study/RiscvPmp/Machine.v
index fb33831..65128d0 100644
--- a/case_study/RiscvPmp/Machine.v
+++ b/case_study/RiscvPmp/Machine.v
@@ -1172,6 +1172,7 @@ Module Import RiscvPmpProgram <: Program RiscvPmpBase.
        | RISCV_SUB => rs1_val - rs2_val
        | RISCV_SRA  => let: tmp := stm_foreign (vector_subrange 4 0) [rs2_val] in
                        call shift_right_arith32 rs1_val tmp
+       | RISCV_ADD' => rs1_val +ᵇ rs2_val
        end in
      call wX rd result ;;
      stm_val ty_retired RETIRE_SUCCESS.
